bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java

```diff
@@ -0,0 +1,370 @@
+ * @author Hans-Jörg Merk - refactored for openHAB 3.0 compatibility
+ *
+ */
+@NonNullByDefault
+public class IntesisBoxHandler extends BaseThingHandler implements IntesisBoxChangeListener {
+
+    private final Logger logger = LoggerFactory.getLogger(IntesisBoxHandler.class);
+    private @Nullable IntesisBoxSocketApi intesisBoxSocketApi;
+
+    private final Map<String, String> properties = new HashMap<>();
+    private final Map<String, List<String>> limits = new HashMap<String, List<String>>();
+
+    private final IntesisDynamicStateDescriptionProvider intesisStateDescriptionProvider;
+
+    private IntesisBoxConfiguration config = new IntesisBoxConfiguration();
+
+    private double minTemp = 0.0, maxTemp = 0.0;
+
+    private @Nullable ScheduledFuture<?> pollingTask;
+
+    public IntesisBoxHandler(Thing thing, IntesisDynamicStateDescriptionProvider intesisStateDescriptionProvider) {
+        super(thing);
+        this.intesisStateDescriptionProvider = intesisStateDescriptionProvider;
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(IntesisBoxConfiguration.class);
+
+        if (!config.ipAddress.isEmpty()) {
+
+            updateStatus(ThingStatus.UNKNOWN);
+            scheduler.submit(() -> {
+                addChannel(CHANNEL_TYPE_AMBIENTTEMP, "Number:Temperature");
+                addChannel(CHANNEL_TYPE_ERRORCODE, "String");
+                addChannel(CHANNEL_TYPE_ERRORSTATUS, "String");
+
+                intesisBoxSocketApi = new IntesisBoxSocketApi(config.ipAddress, config.port);
+                intesisBoxSocketApi.addIntesisBoxChangeListener(this);
+                try {
+                    IntesisBoxSocketApi intesisLocalApi = intesisBoxSocketApi;
+                    if (intesisLocalApi != null) {
+                        intesisLocalApi.openConnection();
+                        intesisLocalApi.sendId();
+                        intesisLocalApi.sendLimitsQuery();
+                        intesisLocalApi.sendAlive();
+                    }
+                } catch (IOException e) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                }
+                pollingTask = scheduler.scheduleWithFixedDelay(this::polling, 3, 45, TimeUnit.SECONDS);
+                updateStatus(ThingStatus.ONLINE);
+            });
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "No IP address specified)");
+        }
+    }
+
+    @Override
+    public void dispose() {
+        final ScheduledFuture<?> pollingTask = this.pollingTask;
+
+        IntesisBoxSocketApi api = this.intesisBoxSocketApi;
+
+        if (pollingTask != null) {
+            pollingTask.cancel(true);
+            this.pollingTask = null;
+        }
+        if (api != null) {
+            api.closeConnection();
+        }
```

Comment:
Is there somewhere here you are calling `intesisBoxSocketApi.removeIntesisBoxChangeListener(this);`? If you aren't then you should.

Reply:
Done